api/buckets/service.go

```diff
@@ -1687,6 +1681,88 @@ func (s *Service) getGatewayHost() (host string, ok bool) {
 	return
 }
 
+func (s *Service) unpinPath(ctx context.Context, path path.Path) error {
+	if err := s.IPFSClient.Pin().Rm(ctx, path); err != nil {
+		return err
+	}
+	node, err := s.IPFSClient.Unixfs().Get(ctx, path)
+	if err != nil {
+		return fmt.Errorf("getting node from path: %s", err)
+	}
+	size, err := node.Size()
+	if err != nil {
+		return fmt.Errorf("getting size of removed node: %s", err)
+	}
+	if err := s.sumBytesPinned(ctx, -size); err != nil {
+		return fmt.Errorf("substracting unpinned node from quota: %s", err)
+	}
+	return nil
+}
+
+// pinBlocks pin the provided blocks to the IPFS node, and accounts to the
+// account/user buckets total size quota.
+func (s *Service) pinBlocks(ctx context.Context, nodes []ipld.Node) error {
+	if err := s.IPFSClient.Dag().Pinning().AddMany(ctx, nodes); err != nil {
+		return fmt.Errorf("pinning set of nodes: %s", err)
+	}
+	var totalAddedSize int64
+	for _, n := range nodes {
+		s, err := n.Size()
+		if err != nil {
+			return fmt.Errorf("getting size of node: %s", err)
+		}
+		totalAddedSize += int64(s)
+	}
+	if err := s.sumBytesPinned(ctx, totalAddedSize); err != nil {
+		return fmt.Errorf("adding pinned size to account quota: %s", err)
+	}
+	return nil
+}
+
+// sumBytesPinned accounts the provided delta difference into the account/user
+// buckets total size quota.
+func (s *Service) sumBytesPinned(ctx context.Context, delta int64) error {
+	key, isUser := ownerFromContext(ctx)
+	if !isUser {
+		a, err := s.Collections.Accounts.Get(ctx, key)
+		if err != nil {
+			return fmt.Errorf("getting account from context: %s", err)
+		}
+		newTotalSize := a.BucketsTotalSize + delta
+		if err := s.Collections.Accounts.SetBucketsTotalSize(ctx, key, newTotalSize); err != nil {
+			return fmt.Errorf("updating new account buckets total size: %s", err)
+		}
+		return nil
+	}
+	u, err := s.Collections.Users.Get(ctx, key)
+	if err != nil {
+		return fmt.Errorf("getting user from context: %s", err)
+	}
+	newTotalSize := u.BucketsTotalSize + delta
+	if err := s.Collections.Users.SetBucketsTotalSize(ctx, key, newTotalSize); err != nil {
+		return fmt.Errorf("updating new users buckets total size: %s", err)
+	}
+	return nil
+}
+
+// sumBytesPinned returns the current buckets total size usage of the account/user
+// logged in the context.
+func (s *Service) getBucketsTotalSize(ctx context.Context) (int64, error) {
+	key, isUser := ownerFromContext(ctx)
+	if !isUser {
+		a, err := s.Collections.Accounts.Get(ctx, key)
```

Comment:
Still seems odd to get the account twice... it's already on the context.

Reply:
Will change. 💯